wpilibc/src/test/native/cpp/DriverStationTest.cpp

```diff
@@ -0,0 +1,83 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2020 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "frc/DriverStation.h"
+#include "frc/Joystick.h"
+#include "frc/simulation/DriverStationSim.h"
+#include "frc/simulation/SimHooks.h"
+#include "frc2/Timer.h"
+#include "gtest/gtest.h"
+
+TEST(DriverStation, NoFMSNoWarnings) {
+  testing::internal::CaptureStderr();
+
+  frc::sim::DriverStationSim::SetFmsAttached(false);
+  frc::sim::DriverStationSim::NotifyNewData();
+  frc::DriverStation::GetInstance().SilenceJoystickConnectionWarnings(true);
+  frc::Joystick joystick{0};
+
+  joystick.GetRawButton(1);
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+  EXPECT_TRUE(frc::DriverStation::GetInstance()
+                  .AreJoystickConnectionWarningsSilenced());
+  EXPECT_EQ(testing::internal::GetCapturedStderr(), "");
+}
+
+TEST(DriverStation, NoFMSWithWarnings) {
+  testing::internal::CaptureStderr();
+
+  frc::sim::DriverStationSim::SetFmsAttached(false);
+  frc::sim::DriverStationSim::NotifyNewData();
+  frc::DriverStation::GetInstance().SilenceJoystickConnectionWarnings(false);
+  frc::Joystick joystick{0};
+
+  joystick.GetRawButton(1);
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+  EXPECT_FALSE(frc::DriverStation::GetInstance()
+                   .AreJoystickConnectionWarningsSilenced());
+  EXPECT_EQ(
+      testing::internal::GetCapturedStderr(),
+      "Joystick Button missing, check if all controllers are plugged in\n");
+}
+
+TEST(DriverStation, WithFMSNoWarnings) {
+  testing::internal::CaptureStderr();
+
+  frc::sim::DriverStationSim::SetFmsAttached(true);
+  frc::sim::DriverStationSim::NotifyNewData();
+  frc::DriverStation::GetInstance().SilenceJoystickConnectionWarnings(true);
+  frc::Joystick joystick{0};
+
+  joystick.GetRawButton(1);
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
```

Comment:
`frc::sim::StepTiming()` can be used to make the message interval pass instead of `std::this_thread::sleep_for()`. I assume the sleep is used to work around the race condition from the print on the separate DS thread.

Reply:
Unfortunately the HAL SendError messages depend on wall clock time so simply stepping through sim timing won't work here.